### PR TITLE
tolerance.json support

### LINKED ISSUE #601: Allow test checker to handle tolerances on per-field basis
It would be nice if we could specify this on a per-field basis, but that could be a follow-on PR.

_Originally posted by @jamiebramwell in https://github.com/LLNL/serac/pull/598#discussion_r755388079_

Options:

* file that has tolerances in it listed: `--tolerances=mytolerances.json`
* parsed comandline string: `--tolerances=velocity:0.1::displacement:0.234`

### PR DESCRIPTION
**Pull request by chapman39**
In `tolerance_test.py`, you can now use a tolerance.json file or one single tolerance value to specify tolerance values for fields. There is also error checking in case there is no default value or no value for each field.

I still need to add things on the `tests/integration` side as well as implementing a string parser for the `--tolerance` option.

fixes #601

### CODE REVIEW

**Comment by chapman39** [comment]
updated tests repo to ensure it works for all tests. you can now specify the tolerance with a string or a json file.

**Comment by chapman39** [comment]
connected with https://github.com/LLNL/serac_tests/pull/4

**Comment by white238** [comment]
@jamiebramwell do we want a string parser for `--tolerance`?  for example, `--tolerance=velocity:0.001,default:0.01`.  I think the default for everything on the command line or a json file is enough personally.

Edit: Ah I see @chapman39 already implemented it. I read the PR description before looking at the file diff...

File: scripts/testing/tolerance_test.py [hunk 1]
```diff
@@ -37,32 +37,54 @@ def ensure_file(path):
         print("ERROR: Given file is not a file: {0}".format(path))
         sys.exit(1)
 
-
 def parse_args():
     parser = argparse.ArgumentParser(description="Compare two Serac summary files")
+    group = parser.add_mutually_exclusive_group(required=True)
```
Comment (white238) [inline scripts/testing/tolerance_test.py]: Can you rename this something that jumps out more like `tolerance_group`?  I like this feature btw, never seen it before.

File: scripts/testing/tolerance_test.py [hunk 2]
```diff
@@ -37,32 +37,54 @@ def ensure_file(path):
...
+    tolerance_dict = {}
+
+    # Two cases:
+    # 1. single value (0.1)
+    # 2. specific value(s) (velocity:0.1::displacement:0.025)
+
+    if ":" not in query:
+        tolerance_dict = { "default": float(query) }
+    else:
+        query_list = query.split("::")
```
Comment (white238) [inline scripts/testing/tolerance_test.py]: @jamiebramwell you originally suggested the delimiter `::`.  Are you okay with a different delimiter to make the field tolerances stand out more, possibly `,`?

`--tolerance=velocity:0.1::displacement:0.234`

vs

`--tolerance=velocity:0.1,displacement:0.234`

File: scripts/testing/tolerance_test.py [hunk 3]
```diff
@@ -37,32 +37,54 @@ def ensure_file(path):
...
     parser = argparse.ArgumentParser(description="Compare two Serac summary files")
+    group = parser.add_mutually_exclusive_group(required=True)
 
     parser.add_argument("--baseline", type=str, required=True,
                         help="Path to baseline summary file")
     parser.add_argument("--test", type=str, required=True,
                         help="Path to test summary file")
-    parser.add_argument("--tolerance", type=float, required=True,
+    group.add_argument("--tolerance", type=str,
                         help="Allowed tolerance amount for individual values")
```
Comment (white238) [inline scripts/testing/tolerance_test.py:68]: Can we add a longer help message that shows possible values here and an example of the JSON file below? You should be able to do it in by passing a multi-line string to line 41.

Something like:

```
  usage = """
  Compare two Serac summary files
  
  Example usages for --tolerance:
  
  --tolerance=0.001 // used for all field values
  --tolerance=default:0.001::velocity:0.1::displacement:0.234  // defaults to 0.001 with other specific tolerances used
  
  Tolerance JSON file example:
  
  {
 "default": 0.0001,  // used as default
 "velocity": 0.001,   // specific field value tolerance
 "displacement":0.01
}
  """

  parser = argparse.ArgumentParser(description=usage)
```

Comment (white238) [reply scripts/testing/tolerance_test.py:68]: Also need a new section here to explain the usage on this page:

https://serac.readthedocs.io/en/latest/sphinx/dev_guide/testing.html

As well as any changes required in this ATS helper function:

https://serac.readthedocs.io/en/latest/sphinx/dev_guide/testing.html#ats-test-helper-functions

**Comment by chapman39** [comment]
it passes the checks, the doc page about testing will be slightly inaccurate until the serac_tests pr goes through, since it relies on `test.ats`.

https://github.com/LLNL/serac_tests/pull/4/

https://serac.readthedocs.io/en/feature-chapman39-unique_tolerances/sphinx/dev_guide/testing.html

**Comment by white238** [comment]
The `tests` submodule needs to point to your new branch.

**Comment by chapman39** [comment]
https://serac.readthedocs.io/en/feature-chapman39-unique_tolerances/sphinx/dev_guide/testing.html

File: scripts/testing/tolerance_test.py [hunk 4]
```diff
@@ -37,32 +37,74 @@ def ensure_file(path):
...
+    usage = """
+Compare two Serac summary files.
+
+Example usages for --tolerance:
+    --tolerance=0.001 // used for all field values
+    --tolerance=default:0.001,velocity:0.1,displacement:0.234 // default is for all non-specified tolerances
+
+Example usage for --tolerance-file:
+    --tolerance-file=tolerance_low.json
+    (Note: <name>.json file must be in tests/integration/tolerances directory)
```
Comment (white238) [inline scripts/testing/tolerance_test.py]: related to comment in other PR: 

https://github.com/LLNL/serac_tests/pull/4/files#r854621606

File: scripts/testing/tolerance_test.py [hunk 5]
```diff
@@ -37,32 +37,74 @@ def ensure_file(path):
...
 
-
 def parse_args():
-    parser = argparse.ArgumentParser(description="Compare two Serac summary files")
+    usage = """
+Compare two Serac summary files.
+
+Example usages for --tolerance:
+    --tolerance=0.001 // used for all field values
+    --tolerance=default:0.001,velocity:0.1,displacement:0.234 // default is for all non-specified tolerances
```
Comment (jamiebramwell) [inline scripts/testing/tolerance_test.py:46]: 👍

**Review by jamiebramwell — APPROVED** [review]
Sorry for the extreme delay. Both the file workflow and the command line args look good!

**Review by white238 — APPROVED** [review]

**Pull request merged**